### PR TITLE
Make Field macro a volatile cast

### PR DESCRIPTION
**Pull request by kayceesrk**
The PR addresses the issue with C FFI and the OCaml memory model. For context see, #10992. 

The first commit make `Field` macro a volatile cast, and the second commit introduces `Load_field` macro that is not an l-value and performs an acquire load of the given field. 

## Goal

The comment at https://github.com/ocaml/ocaml/issues/10992#issuecomment-1112394309 gives a detailed description of the aim of the PR. I shall try to summarize the issue and the aim of the PR. 

The PR https://github.com/ocaml/ocaml/pull/10972 performs a compilation of the OCaml memory model to C11 (`caml_modify`, `caml_atomic_*`, etc) and hardware models (x86 and ARMv8). This compilation covers all of the FFI functions except `Field`, which is the focus on #10992. `Field` macro on trunk performs non-atomic reads. Since this is C, data races using the `Field` macro will have catch fire semantics. 

Can we make `Field` macro an `_Atomic` cast to avoid catch-fire semantics? No. `Field` macro is also an l-value and making `Field` and `_Atomic` cast will make it not be an l-value. Our previous experience with the concurrent major GC shows that making `Field` not an l-value breaks an awful lot of packages. Hence, we make `Field` a volatile cast, which, in practice, avoids catch-fire semantics. See Linux kernel memory model https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p0124r7.html#Variable%20Access. 

We also introduce `Load_field` macro which does an acquire load. The aim of the `Load_field` macro is to provide a mechanism for users to read fields in C11 compatible fashion  (and not "in practice" assumptions that comes with `volatile` fields) that also preserves the ordering necessary for OCaml memory model. We shall encourage our users to start using `Load_field` in places where `Field` is used as an r-value. Technically, `Load_field` is only necessary for mutable fields, and immutable fields can be accessed using `Field` macro. However, for users, it will be easier to switch to `Load_field` unconditionally. 

## Possible backwards compatibility concern

Given that `Field` has been made volatile, the uses of `&Field(..)` will have the type `volatile value *`. This will cause C compiler to raise a warning about discarded quantifiers when functions that take a `value *` argument are passed `volatile value *` value.

```c
error: passing argument 1 of ‘foo’ discards ‘volatile’ qualifier from pointer target type [-Werror=discarded-qualifiers]
```

The PR includes changes to the types of certain functions (`caml_modify`, `caml_initialize`, ...) to make the C compiler happy. 

## Todo

- [x] Performance benchmarking -- Run Sandmark on this branch to measure the impact of the change.
- [x] Assess the impact of backwards `&Field(..)` using opam health check. 

CC @gadmm @stedolan.

### CODE REVIEW

**Comment by kayceesrk** [comment]
image

The graph above shows the normalized running time for this PR against trunk on macro benchmarks from Sandmark. The results show that there is neither a definitive slowdown nor a speedup overall; there's a max 4% swing in both directions. Given that this PR is necessary for correctness, I would consider that there is nothing to be done on the performance side.

**Comment by kayceesrk** [comment]
@dra27 ran opam health check on this PR as well as this patch applied to 4.14. For 5.0.0+trunk, there are no new breakages. Every package that builds on trunk builds with this PR. For 4.14, the only breakage is `ocamlnet`. `ocamlnet` is broken on 5.0.0 already. Thanks, @dra27 for running opam health check.

**Comment by gasche** [comment]
The Sandmark results are a bit weird, as we don't see how the change could possibly improve performance. (Using volatile is supposed to disable some compiler optimizations only.) This suggests that the changes we observe, at least on the "speed up" part of the graph, are mostly tooling-related noise (maybe the different code generation creates weird code-placement effects, or maybe it is actually faster in some case to force the compiler to recompute reads, etc.). Given that the "slow down" parts seem comparable in magnitude to the "speed up" part, it's hard to tell if the results are actually meaningful.

Said otherwise: for a change that should not impact performance at all, maybe what we should expect to observe in practice, because life is complicated, is for speedup/slowdown results to fall along a gaussian distribution centered at 0. I wonder whether the results here fit this model.

**Comment by kayceesrk** [comment]
I analyzed the outliers. The observed results are reproducible and not an artifact of sandmark. 

## `test_decrompress` (max speedup)

### Hyperfine

```bash
kc@kaveri:~/repos/sandmark/_build$ hyperfine './5.0.0+trunk_1/benchmarks/decompress/test_decompress.exe 64 524_288' './5.0.0+trunk+volatile_field_1/benchmarks/decompress/test_decompress.exe 64 524_288'
Benchmark #1: ./5.0.0+trunk_1/benchmarks/decompress/test_decompress.exe 64 524_288
  Time (mean ± σ):      3.189 s ±  0.041 s    [User: 3.179 s, System: 0.009 s]
  Range (min … max):    3.130 s …  3.274 s    10 runs
 
Benchmark #2: ./5.0.0+trunk+volatile_field_1/benchmarks/decompress/test_decompress.exe 64 524_288
  Time (mean ± σ):      3.060 s ±  0.028 s    [User: 3.048 s, System: 0.012 s]
  Range (min … max):    3.012 s …  3.090 s    10 runs
 
Summary
  './5.0.0+trunk+volatile_field_1/benchmarks/decompress/test_decompress.exe 64 524_288' ran
    1.04 ± 0.02 times faster than './5.0.0+trunk_1/benchmarks/decompress/test_decompress.exe 64 524_288'
```

### Perf 

Here are the perf results:

trunk:
```bash
kc@kaveri:~/repos/sandmark/_build$ perf stat ./5.0.0+trunk_1/benchmarks/decompress/test_decompress.exe 64 524_288

 Performance counter stats for './5.0.0+trunk_1/benchmarks/decompress/test_decompress.exe 64 524_288':

          3,148.70 msec task-clock                #    1.000 CPUs utilized          
                 5      context-switches          #    0.002 K/sec                  
                 0      cpu-migrations            #    0.000 K/sec                  
             5,265      page-faults               #    0.002 M/sec                  
     9,240,551,815      cycles                    #    2.935 GHz                    
    28,028,690,152      instructions              #    3.03  insn per cycle         
     5,057,837,328      branches                  # 1606.328 M/sec                  
        10,103,793      branch-misses             #    0.20% of all branches        

       3.149228619 seconds time elapsed

       3.145162000 seconds user
       0.004001000 seconds sys
```

this PR:
```bash
kc@kaveri:~/repos/sandmark/_build$ perf stat ./5.0.0+trunk+volatile_field_1/benchmarks/decompress/test_decompress.exe 64 524_288

 Performance counter stats for './5.0.0+trunk+volatile_field_1/benchmarks/decompress/test_decompress.exe 64 524_288':

          3,139.49 msec task-clock                #    1.000 CPUs utilized          
                 9      context-switches          #    0.003 K/sec                  
                 0      cpu-migrations            #    0.000 K/sec                  
             7,013      page-faults               #    0.002 M/sec                  
     9,136,767,429      cycles                    #    2.910 GHz                    
    28,114,813,808      instructions              #    3.08  insn per cycle         
     5,058,882,903      branches                  # 1611.370 M/sec                  
        10,492,631      branch-misses             #    0.21% of all branches        

       3.140096681 seconds time elapsed

       3.123928000 seconds user
       0.015999000 seconds sys
```

This PR seems to cause 0.3% more instructions to be executed (28.114B vs 28.028B), due to missed compiler optimisations(?). But those instructions are executed faster (3.08 insn per cycle vs 3.03 insn per cycle).

## kb (max slowdown)

### Hyperfine

```bash
kc@kaveri:~/repos/sandmark/_build$ hyperfine './5.0.0+trunk_1/benchmarks/kb/kb.exe' './5.0.0+trunk+volatile_field_1/benchmarks/kb/kb.exe'
Benchmark #1: ./5.0.0+trunk_1/benchmarks/kb/kb.exe
  Time (mean ± σ):      2.832 s ±  0.018 s    [User: 2.825 s, System: 0.007 s]
  Range (min … max):    2.816 s …  2.866 s    10 runs
 
Benchmark #2: ./5.0.0+trunk+volatile_field_1/benchmarks/kb/kb.exe
  Time (mean ± σ):      2.915 s ±  0.036 s    [User: 2.909 s, System: 0.006 s]
  Range (min … max):    2.852 s …  2.979 s    10 runs
 
Summary
  './5.0.0+trunk_1/benchmarks/kb/kb.exe' ran
    1.03 ± 0.01 times faster than './5.0.0+trunk+volatile_field_1/benchmarks/kb/kb.exe'
```

### Perf

trunk:
```
kc@kaveri:~/repos/sandmark/_build$ perf stat ./5.0.0+trunk_1/benchmarks/kb/kb.exe

 Performance counter stats for './5.0.0+trunk_1/benchmarks/kb/kb.exe':

          2,810.43 msec task-clock                #    1.000 CPUs utilized          
                 4      context-switches          #    0.001 K/sec                  
                 0      cpu-migrations            #    0.000 K/sec                  
             2,137      page-faults               #    0.760 K/sec                  
     8,244,770,572      cycles                    #    2.934 GHz                    
    16,000,114,884      instructions              #    1.94  insn per cycle         
     3,626,059,322      branches                  # 1290.216 M/sec                  
       141,430,823      branch-misses             #    3.90% of all branches        

       2.810932906 seconds time elapsed

       2.810875000 seconds user
       0.000000000 seconds sys
```

this PR:
```bash
kc@kaveri:~/repos/sandmark/_build$ perf stat ./5.0.0+trunk+volatile_field_1/benchmarks/kb/kb.exe

 Performance counter stats for './5.0.0+trunk+volatile_field_1/benchmarks/kb/kb.exe':

          2,943.68 msec task-clock                #    1.000 CPUs utilized          
                 4      context-switches          #    0.001 K/sec                  
                 0      cpu-migrations            #    0.000 K/sec                  
             2,134      page-faults               #    0.725 K/sec                  
     8,682,540,315      cycles                    #    2.950 GHz                    
    16,015,244,224      instructions              #    1.84  insn per cycle         
     3,625,079,871      branches                  # 1231.477 M/sec                  
       148,965,798      branch-misses             #    4.11% of all branches        

       2.944208362 seconds time elapsed

       2.936149000 seconds user
       0.008000000 seconds sys
```

This PR causes 0.09% more instructions to be executed. But they're executed at a slower rate than trunk (1.94 vs 1.84 insn per cycle).

## Observation

We can see that the results are not due to measurement noise but are observable. But I'm not sure whether these are actionable since this looks like micro-architectural behaviours at play.

**Comment by gasche** [comment]
I would say that there are different sorts of noise. Some sources of noise generate variance from one run of the benchmark to the other (and the benchmark setup is trying to eliminate it), or from one machine to the next. But there are other sorts of noise that still qualify as noise, because they are small changes in result that do not come from a fundamental change in runtime behavior, but unpredictable complex systems placed between our source code and the behaviors we observe (for example, CPUs and their micro-architectural effects).

That latter source of "noise" (which I called "tooling noise") is what seems to be going on here. In this noise category, the performance difference that we observe on one benchmark is a deterministic function of the benchmark, but the effect on a family of benchmarks (the function from a benchmark to the performance difference) could be random/arbitrary. Maybe we can model the impact of this change as a random speedup/slowdown for each benchmark, following a Gaussian centered at 0 -- many programs seeing their performance unchanged, and a small number of benchmarks seeing small speedups or slowdowns. (Or maybe there is a mix of "micro-architectural effects with winners and losers" and actual real slowdowns, caused by say a specific runtime function being pessimized; in that case the average speedup/slowdown would be tilted towards the "slowdown" part, but if such an effect exists it is not clear in the benchmark suite.)

In summary: we observe speedups of 4% that have no clear reason (possibly: obscure microarchitectural effects), which suggests that we can discard all performance effects of up to 4% as just weird magic. Only much larger changes would be representative, and there are none in this benchmark suite.

**Comment by kayceesrk** [comment]
Thanks for the summary. In the past, we have had features that introduce swings of 2-4% in both directions. A good approximation of the overall change is the geometric mean of the normalised running times. For this PR, the geomean is `0.9998`. That is, with this PR, the benchmarks run very marginally faster than trunk (by 0.02%).

**Comment by Octachron** [comment]
A possible testable null-hypothesis is that the change is no more likely to decrease performance than to increase them.
Thus, if we consider the number of test with decreased performance minus `D` after `N` i.i.d test, its distribution should be a binomial:
```ocaml
P(D = d) = (1/2)^N N!/((N-D)!D!
```
With 50 samples, this hypothesis can be rejected with a p-value of 5% if we observe 31 samples with decreased performances. Since this is not the case, this null hypothesis is not observably wrong.

For testing a gaussian hypothesis, I would need to know the variance. @kayceesrk , if you send me the raw data I can have a look. However, I am not sure that the extremal values observed in the benchmark are compatible with a gaussian hypothesis: with 50 samples, it is possible to reject a gaussian hypothesis with 5% p-value if the maximum is greater than 3.08 σ and that might be the case here.

**Comment by gasche** [comment]
It's surprisingly hard to find clear explanation on this problem that looks common, important and affecting people who are not experts in benchmark design. (What's the expected distribution of the performance impact of microarchitectural effects these days?) All I could find was [STABILIZER: statistically sound performance evaluation, Curtsinger and Berger, 2013](https://dl.acm.org/doi/10.1145/2451116.2451141), a paper on a complex system to dynamically randomize the (code, heap, stack) layout of running programs, performed using instrumentation at the LLVM level. (It looks like it would be quite some amount of work to reproduce it in the OCaml compiler, but we could actually try to enable this on the OCaml *runtime code*.) They report that, using their randomization tool, they observe that layout effects follow a Gaussian distribution ("very often"; they don't on two specific benchmarks).

**Comment by xavierleroy** [comment]
Maybe we could have this general discussion (on interpreting Sandmark results) separately from this PR.  But, yes, differences in code placement can affect execution times by several percents.  To evaluate the impact of optimizations, I prefer to count the number of instructions executed, there's less noise this way.

File: runtime/array.c [hunk 1]
```diff
@@ -583,7 +584,7 @@ CAMLprim value caml_array_fill(value array,
         caml_darken(NULL, old, NULL);
       }
       if (is_val_young_block)
-        Ref_table_add(&Caml_state->minor_tables->major_ref, fp);
+        Ref_table_add(&Caml_state->minor_tables->major_ref, (value*)fp);
```
Comment (xavierleroy) [inline runtime/array.c]: I think you don't need the `(value *)` cast, since `Ref_table_add` contains a cast already.

File: runtime/compare.c [hunk 2]
```diff
@@ -280,8 +280,8 @@ static intnat do_compare_val(struct compare_stack* stk,
       if (sz1 > 1) {
         sp++;
         if (sp >= stk->limit) sp = compare_resize_stack(stk, sp);
-        sp->v1 = &Field(v1, 1);
-        sp->v2 = &Field(v2, 1);
+        sp->v1 = (value*)&Field(v1, 1);
+        sp->v2 = (value*)&Field(v2, 1);
```
Comment (xavierleroy) [inline runtime/compare.c]: I'd rather change the declaration of `struct compare_item`:
```
struct compare_item { volatile value * v1, * v2; mlsize_t count; };
```

File: runtime/extern.c [hunk 3]
```diff
@@ -845,7 +845,7 @@ static void extern_rec(struct caml_extern_state* s, value v)
         sp++;
         if (sp >= s->extern_stack_limit)
           sp = extern_resize_stack(s, sp);
-        sp->v = &Field(v, i + 1);
+        sp->v = (value*)&Field(v, i + 1);
```
Comment (xavierleroy) [inline runtime/extern.c]: Likewise, I'd rather change the definition of `struct extern_item`:
```
struct extern_item { volatile value * v; mlsize_t count; };
```

File: runtime/intern.c [hunk 4]
```diff
@@ -328,7 +328,7 @@ static struct intern_item * intern_resize_stack(struct caml_intern_state* s,
     if (_n > 0) {                                                       \
       PushItem(s);                                                      \
       sp->op = OReadItems;                                              \
-      sp->dest = _dest;                                                 \
+      sp->dest = (value*)_dest;                                         \
```
Comment (xavierleroy) [inline runtime/intern.c]: Along the same lines, you could make `dest` volatile in `struct intern_item`, and declare the second argument of `intern_rec` volatile.

File: runtime/interp.c [hunk 5]
```diff
@@ -638,21 +638,21 @@ value caml_interprete(code_t prog, asize_t prog_size)
       if (nvars > 0) *--sp = accu;
       if (blksize <= Max_young_wosize) {
         Alloc_small(accu, blksize, Closure_tag, Enter_gc);
-        p = &Field(accu, envofs);
+        p = (value*)&Field(accu, envofs);
```
Comment (xavierleroy) [inline runtime/interp.c]: Likewise, what about declaring `p` as `volatile value *` ?

File: runtime/minor_gc.c [hunk 6]
```diff
@@ -413,7 +413,7 @@ static void oldify_mopup (struct oldify_state* st, int do_ephemerons)
          re < ephe_ref_table.ptr; re++) {
       value *data = re->offset == CAML_EPHE_DATA_OFFSET
               ? &Ephe_data(re->ephe)
-              :  &Field(re->ephe, re->offset);
+              : (value*)&Field(re->ephe, re->offset);
```
Comment (xavierleroy) [inline runtime/minor_gc.c]: Here you could avoid the cast by declaring `volatile value * data`.  There would be an implicit conversion from non-volatile to volatile in the `Ephe_data` branch.

**Review by xavierleroy — COMMENTED** [review]
There's a few extra points in the code where you could avoid a `(value *)` cast by declaring some variables or struct fields `volatile value *`.  On the one hand, I like it when there are fewer casts.  On the other hand those `volatile` declarations add noise.  I don't feel strongly either way.

File: runtime/minor_gc.c (hunk 6 above)
Comment (kayceesrk) [reply runtime/minor_gc.c]: This change necessitated changing the type of the `value*` argument in `oldify_one` to `volatile value*`.

**Comment by kayceesrk** [comment]
Building upon @xavierleroy's comments, I refactored the PR further to remove almost all of the casts of `volatile value*` to `value*` by declaring variables as `volatile value*`. The code feels easier to reason about than casting to remove the `volatile` qualifier. 

The only two places where we explicitly cast `volatile value*` to `value*` are 

1. In the call to `memmove`, where the type cast is necessary.
2. In `Ref_table_add`. All ref tables are local to a particular domain.

**Review by gasche — APPROVED** [review]
I like the new patchset, it's clean, and the rationale for the change has been discussed at length. We haven't been able to detect a performance regression in the benchmarks, as we expected.

(I wondered if it would make sense to introduce an alias for `volatile value *`, maybe `lvalue` :-' but this is no priority.)

**Comment by gasche** [comment]
I think that a Changes entry would make sense: this is a user-visible change, including compared to 4.x, and it would be useful to provide clear context for it and also to credit Guillaume's contribution.

File: runtime/minor_gc.c (hunk 6 above)
Comment (xavierleroy) [reply runtime/minor_gc.c]: Looks like a reasonable change to me.

**Review by xavierleroy — APPROVED** [review]
Looks very good to me.  Thanks!

**Comment by gadmm** [comment]
I would like to have a look at this PR, it seems like I will have various comments. Can you please leave me the time to review it later?

In the meanwhile, I have not seen much discussion about the proposal at https://github.com/ocaml/ocaml/issues/10992#issuecomment-1112394309 which motivates this PR. Would some memory-model expert who feel involved with OCaml please tell us what they think about it?

**Comment by kayceesrk** [comment]
> Can you please leave me the time to review it later?

Sounds good. Thanks for having a look. I'll wait until the review is completed to write the Changes entry.

**Comment by xavierleroy** [comment]
We need to make progress towards the 5.0 release, so @gadmm please review ASAP or tell us you can't do it, but don't make us wait.

**Comment by gadmm** [comment]
I'll give it a try over the week-end.

**Comment by gadmm** [comment]
> The PR addresses the issue with C FFI and the OCaml memory model. For context see, #10992.

To be clearer, this PR does not close #10992 by itself.

> The first commit make `Field` macro a volatile cast, and the second commit introduces `Load_field` macro that is not an l-value and performs an acquire load of the given field.

I would like that `Load_field` is added in a separate PR. First, the proposal is to add an acquire fence when going from OCaml to C to reflect losing track of data dependencies, so that an expert user does not have to use `Load_field` for immutable fields. It will be the occasion to discuss my reasoning or tell me that it is not necessary (the counter-example is a mutable load operated on the OCaml side followed by two nested immutable loads on the C side, maybe one is enough). Adding the fence involves changes to the .S files, and perhaps benchmarking the change (e.g. do you also add it to `noalloc` calls or is it too costly?). Second, it would be good if the PR could include its documentation so we can be sure that we agree on its recommended usage.

> The PR #10972 performs a compilation of the OCaml memory model to C11 (`caml_modify`, `caml_atomic_*`, etc) and hardware models (x86 and ARMv8). This compilation covers all of the FFI functions except `Field`, which is the focus on #10992. 

Does this PR include a review that there are no other macros similar to `Field`?

> Can we make `Field` macro an `_Atomic` cast to avoid catch-fire semantics? No. `Field` macro is also an l-value and making `Field` and `_Atomic` cast will make it not be an l-value. Our previous experience with the concurrent major GC shows that making `Field` not an l-value breaks an awful lot of packages.

With `#define Field(x, i) (((_Atomic(value) *)(x)) [i])` it can be used both as an l- and r-value. However, this translates into SC-store/loads, which have a prohibitive cost on all platforms. It would however avoid catch-fire semantics both in practice and theory unlike `volatile`. I agree that what dooms it is the too great effort needed for users to replace all the occurrences of `Field` in existing code (including initializing stores) by something more sensible (even progressively).

> Given that `Field` has been made volatile, the uses of `&Field(..)` will have the type `volatile value *`. This will cause C compiler to raise a warning about discarded quantifiers when functions that take a `value *` argument are passed `volatile value *` value.
> 
> ```c
> error: passing argument 1 of ‘foo’ discards ‘volatile’ qualifier from pointer target type [-Werror=discarded-qualifiers]
> ```
> 
> The PR includes changes to the types of certain functions (`caml_modify`, `caml_initialize`, ...) to make the C compiler happy.

This is a bigger change than I expected.

First, I think that `volatile` should remain an implementation detail since it is not an official part of the memory model in our proposal (`Load_field` is, instead). So I would prefer if it is hidden for users behind a typedef.

But I am also worried that changing the signature of `caml_modify` and `caml_initialize` simply moves the problem around. Opam health check can be useful to find examples of how code can break, but by definition it cannot show that no code breaks. Here the lesson is you have found in OCaml's own code an example of how code can break (which does not look very specific to code that would only be used internally).

This, together with the risks of users avoiding `Load_field` altogether, makes me rethink the opportunity of adding a `volatile` cast to the `Field` macro, instead of nothing at all.

> Todo
> 
> [x] Performance benchmarking -- Run Sandmark on this branch to measure the impact of the change.

I was wondering (here and for other multicore benchmarks that show similar fluctuations) if you have tried to vary the compiler and the compilation options. Some compilation options create a different code layout, which can sometimes be enough to convince whether the performance swing is relevant or not. (In terms of Florian's qualitative criterion, it gives you more independent data points.)

Can you please point me at the specs of the machine?

> [x] Assess the impact of backwards &Field(..) using opam health check.

Does opam health check catch new warnings, or does it only catch them in the cases where the warning is turned into an error?

**Comment by kayceesrk** [comment]
> I would like that Load_field is added in a separate PR. 

Sure. It would be a good chance to include the documentation for `Load_field` in that PR. 

> Does this PR include a review that there are no other macros similar to Field?

No. 

> With `#define Field(x, i) (((_Atomic(value) *)(x)) [i])` it can be used both as an l- and r-value. However, this translates into SC-store/loads, which have a prohibitive cost on all platforms. It would however avoid catch-fire semantics both in practice and theory unlike volatile.

Good point. Thanks for clarifying. 

Ideally, it would have been better than volatile cast if we were able to write `#define Field(x, i) (((_Atomic(relaxed) value *)(x)) [i])` which treats the loads and stores as having relaxed ordering. Such a facility does not exist in C11. 

> Does opam health check catch new warnings, or does it only catch them in the cases where the warning is turned into an error?

Right. The runtime is built `-Werror`, which is why the error

```bash
error: passing argument 1 of ‘foo’ discards ‘volatile’ qualifier from pointer target type [-Werror=discarded-qualifiers]
```

shows up. Unless the C parts of the libraries are also built with `-Werror`, the build will not error out. @dra27 Does `dune` based compilation of C libraries already do `-Werror`? 

If not, I wonder whether we can do something simple like:

```bash
alias gcc="gcc -Werror"
```

to get an assessment of the failing packages. 

> I was wondering (here and for other multicore benchmarks that show similar fluctuations) if you have tried to vary the compiler and the compilation options. Some compilation options create a different code layout, which can sometimes be enough to convince whether the performance swing is relevant or not. (In terms of Florian's qualitative criterion, it gives you more independent data points.)

No. We don't currently perform benchmarking with changing compiler and compiler options regularly. We will see fluctuations by changing the compilation option. However, it is hard to translate the observed fluctuations into actionable information. That said, what has been useful is to use the sandmark results as a starting point to analyse the performance of outliers in detail as I have done in https://github.com/ocaml/ocaml/pull/11255#issuecomment-1126693075. Firstly, in addition to looking at the running time, the change in the number of instructions executed is useful to look at. But this can also throw surprising results. For the program which runs fastest with this PR compared to trunk (`test_decompress`), the program actually executes 0.3% more instructions. The CPU happens to run them faster. We already have enough fluctuations even without changing the compiler options due to micro-architectural optimizations. 

From my perspective, this PR is needed for correctness. Since the performance impact is not significant, I would consider the performance impact to be acceptable. 

> Can you please point me at the specs of the machine?

https://sandmark.tarides.com/#machines-used-for-generating-results

> First, I think that volatile should remain an implementation detail since it is not an official part of the memory model in our proposal (Load_field is, instead). So I would prefer if it is hidden for users behind a typedef.

Are you thinking about something like `typedef volatile value* valueptr`?

**Comment by xavierleroy** [comment]
> >  The PR includes changes to the types of certain functions (caml_modify, caml_initialize, ...) to make the C compiler happy.

> This is a bigger change than I expected.

From an API user's perspective, the big visible change  is that `&Field(x,i)` is no longer of `value *` type.  Adding a `volatile` modifier to the first  argument of `caml_modify`  is transparent, since you can still pass a plain `value *`.

This proposal does make the assumption that few third-party codes take pointers inside blocks `&Field(x,i)` except to pass them to `caml_modify` or `caml_initialize`.  Indeed, codes that do this have a high chance of breaking the GC interface.  It would be interesting to see what uses can be found in the wild.

Assuming third-party code should better not work with pointers inside blocks, there's no need for a typedef.  I like the verbosity of `volatile value *`: it reminds us that something fishy is going on :-)

**Comment by dra27** [comment]
Dune doesn't add `-Werror` for C files, no - that's only done for OCaml files in the dev profile. Projects can opt-in to that by adding `(env (dev (c_flags :standard -Werror)))` to `dune` files.

However, we don't do this in OCaml's build system either - for a release tag / tarball we also compile C objects _without_ `-Werror`. That's point of doing this change with `volatile`, though - single-domain code continues to compile, just with additional warnings, developers should _develop_ with `-Werror` turned on, and can then address poor/incorrect uses of `Field` in subsequent releases.

I don't think there's a reliable way to turn on `-Werror` for all packages to do a health-check run, but I'm not sure that it's worth it (unless I'm still missing something)?

**Comment by kayceesrk** [comment]
Thanks for the response @dra27. 

I was under the false assumption that the lack of new failures with this PR indicated that third-party code was not taking internal pointers to blocks and using them in interesting ways. Like @xavierleroy, it would be interesting to see whether this pattern does occur in the wild, and how many packages are broken due to this change. That said, I don't think that the result of such an analysis should block this PR.

**Comment by gadmm** [comment]
It would be nice to have a look if any other macro needs the volatile cast so they are all fixed at once (`Field` was only meant as an example of problem).

Yes I am thinking about something like `typedef volatile value * value_ref`. This might depend on usages of `value *` in the wild. I am not convinced that the API discourages the use of pointers. In fact, even OCaml uses value pointers in its API (see `caml_named_value`, which incidentally should certainly return `const volatile value *` instead of `const value *` given that parallel mutations are possible via `caml_register_named_value`). There is also the question of whether people who use `value *` are encouraged or not to switch entirely to `volatile value *` when pointing inside the OCaml heap (if they are careful, they only need to use `Load_field` in the right places).

I agree that adding `volatile` to caml_modify/initialize is better than merely moving the problem, for reasons said.

**Comment by Octachron** [comment]
Am I reading correctly that every one agree that this PR is a good compromise and a step forward? If so, I would propose to first merge this PR, and then move to reviewing other macros in subsequent PRs or issues.

**Comment by xavierleroy** [comment]
I agree this PR is in good shape.  I just rebased to fix conflicts and added a Changes entry.  Will run CI precheck next.

**Comment by xavierleroy** [comment]
CI precheck is positive.  Let's merge!

**Comment by xavierleroy** [comment]
I'll wait for more CI to run on trunk, then pick to 5.0.

**Comment by kayceesrk** [comment]
Thanks for pushing the PR across the line @Octachron and @xavierleroy.

**Comment by xavierleroy** [comment]
CI looks globally good, so here is the 5.0 cherry-pick: 7b76f07aa6

**Comment by gadmm** [comment]
I opened an issue with my comment left without reply at #11426.

File: runtime/caml/mlvalues.h [hunk 7]
```diff
@@ -205,7 +205,7 @@ bits  63        (64-P) (63-P)        10 9     8 7   0
 #define Op_val(x) ((value *) (x))
 #define Op_atomic_val(x) ((atomic_value *) (x))
 /* Fields are numbered from 0. */
-#define Field(x, i) (((value *)(x)) [i])           /* Also an l-value. */
+#define Field(x, i) (((volatile value *)(x)) [i]) /* Also an l-value. */
```
Comment (anmonteiro) [inline runtime/caml/mlvalues.h:208]: `Field(x, 0)` and `Op_val(x)` previously had the exact same behavior, but `Field` is now a volatile cast whereas `Op_val` went unchanged in this PR.

Was this on purpose?

Comment (anmonteiro) [reply runtime/caml/mlvalues.h:208]: I had missed it, but perhaps https://github.com/ocaml/ocaml/issues/11426 will take care of that.
